pkg/api/vm/store.go

```diff
@@ -0,0 +1,173 @@
+package vm
+
+import (
+	"fmt"
+
+	"github.com/rancher/apiserver/pkg/apierror"
+	"github.com/rancher/apiserver/pkg/types"
+	"github.com/rancher/wrangler/pkg/schemas/validation"
+	"github.com/rancher/wrangler/pkg/slice"
+	k8sapierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	ctlcdiv1beta1 "github.com/rancher/harvester/pkg/generated/controllers/cdi.kubevirt.io/v1beta1"
+	ctlkubevirtv1alpha3 "github.com/rancher/harvester/pkg/generated/controllers/kubevirt.io/v1alpha3"
+	ctlcorev1 "github.com/rancher/wrangler-api/pkg/generated/controllers/core/v1"
+)
+
+type vmStore struct {
+	types.Store
+
+	vmCache         ctlkubevirtv1alpha3.VirtualMachineCache
+	dataVolumes     ctlcdiv1beta1.DataVolumeClient
+	configmaps      ctlcorev1.ConfigMapClient
+	secrets         ctlcorev1.SecretClient
+	serviceAccounts ctlcorev1.ServiceAccountClient
+	pvcs            ctlcorev1.PersistentVolumeClaimClient
+}
+
+func (s *vmStore) Delete(request *types.APIRequest, schema *types.APISchema, id string) (types.APIObject, error) {
+	deleteDisks := request.Query["deleteDisks"]
+	vm, err := s.vmCache.Get(request.Namespace, request.Name)
+	if err != nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.ServerError, fmt.Sprintf("Failed to get vm %s/%s, %v", request.Namespace, request.Name, err))
+	}
+
+	var keepDisks []string
+	var keepDataVolumes []string
+	var deletePvc []string
+	var deleteConfigmap []string
+	var deleteSecret []string
+	var deleteServiceAccount []string
+	var deleteDataVolume []string
+	if vm.Spec.Template != nil {
+		for _, v := range vm.Spec.Template.Spec.Domain.Devices.Disks {
+			if !slice.ContainsString(deleteDisks, v.Name) {
+				keepDisks = append(keepDisks, v.Name)
+			}
+		}
+
+		for _, vol := range vm.Spec.Template.Spec.Volumes {
+			if slice.ContainsString(keepDisks, vol.Name) && vol.DataVolume != nil {
+				keepDataVolumes = append(keepDataVolumes, vol.DataVolume.Name)
+			}
+
+			if slice.ContainsString(deleteDisks, vol.Name) {
+				switch {
+				case vol.PersistentVolumeClaim != nil:
+					deletePvc = append(deletePvc, vol.PersistentVolumeClaim.ClaimName)
+				case vol.Secret != nil:
+					deleteSecret = append(deleteSecret, vol.Secret.SecretName)
+				case vol.ConfigMap != nil:
+					deleteConfigmap = append(deleteConfigmap, vol.ConfigMap.Name)
+				case vol.ServiceAccount != nil:
+					deleteServiceAccount = append(deleteServiceAccount, vol.ServiceAccount.ServiceAccountName)
```

Comment:
How do we tell users what we are gonna do when they select these disks? Given that these are not volumes

Reply:
Will support later if we need, remove pvc secret, configmap, serviceAccount manage now.